File: language_construct_performance/bun/scripts/value_obj_fixed_properties.ts
import { NUM_ENTRIES, benchmarkStats, saveStats } from "./benchmark_utils";

// --- VALUE OBJECT CLASSES ---

class AgSType {
  t: number;
  constructor(t: number) {
    this.t = t;
  }
}

class AgASubSType {
  t: number;
  s: number;
  r: number;
  constructor(t: number, s: number, r: number) {
    this.t = t;
    this.s = s;
    this.r = r;
  }
}

class AgAType {
  t: number;
  s: AgASubSType;
  constructor(t: number, s: AgASubSType) {
    this.t = t;
    this.s = s;
  }
}

class AgType {
  s: AgSType;
  a: AgAType;
  constructor(s: AgSType, a: AgAType) {
    this.s = s;
    this.a = a;
  }
}

class DetailsType {
  provider: string;
  accountId: string;
  principal: boolean;
  tags: string[];
  mfas: string;
  la: number;
  ut: number;
  s: number;
  cpd: number;
  pcb: string;
  lld: number;
  cd: number;
  cb: string;
  ub: string;
  ud: number;
  ua: number;
  ag: AgType;

  constructor(
    provider: string, accountId: string, principal: boolean, tags: string[],
    mfas: string, la: number, ut: number, s: number, cpd: number, pcb: string,
    lld: number, cd: number, cb: string, ub: string, ud: number, ua: number, ag: AgType
  ) {
    this.provider = provider; this.accountId = accountId; this.principal = principal;
    this.tags = tags; this.mfas = mfas; this.la = la; this.ut = ut; this.s = s;
    this.cpd = cpd; this.pcb = pcb; this.lld = lld; this.cd = cd; this.cb = cb;
    this.ub = ub; this.ud = ud; this.ua = ua; this.ag = ag;
  }
}

class ValueObjectNode {
  label: string;
  id: string;
  edgeTo: string[];
  accessTo: string[];
  details: DetailsType;

  constructor(label: string, id: string, edgeTo: string[], accessTo: string[], details: DetailsType) {
    this.label = label;
    this.id = id;
    this.edgeTo = edgeTo;
    this.accessTo = accessTo;
    this.details = details;
  }
}

// --- FACTORIES ---

function createValueObject(index: number) {
  return new ValueObjectNode(
    "user-dev-test-" + index,
    "u." + (16406 + index),
    ["r.392", "r.40", "r.41", "update", "administrator", "create", "delete", "read"],
    ["s.[s3].UACDR", "a.[s3].DARC", "s.[secretsmanager].RACDU", "s.[dynamodb].RCDAU"],
    new DetailsType(
      "aws", "568709751681", true, ["aKIAYI2NaRQPOT", "dev testing local"], "",
      1772454942 + (index % 1000), 2, 1, 0, "-", 0, 1763097939000, "-", "-", 0, 1772526871591,
      new AgType(
        new AgSType(167),
        new AgAType(
          3187978,
          new AgASubSType(3187978, 3149311, 42506)
        )
      )
    )
  );
}

// --- BENCHMARK RUNNER ---

const stats: Record<string, any> = {};

async function runBenchmark() {
  console.log(`Starting Value Object Fixed Properties Benchmark with ${NUM_ENTRIES.toLocaleString()} entries...`);

  let valueObjArray: ValueObjectNode[] | null = new Array(NUM_ENTRIES);

  console.log("\n--- CREATION ---");
  benchmarkStats("Value Object Creation", stats, "creationTimeMs", () => {
    for (let i = 0; i < NUM_ENTRIES; i++) {
      valueObjArray![i] = createValueObject(i);
    }
  }, true);

  console.log("\n--- PLAIN TRAVERSAL ---");
  benchmarkStats("Plain Traversal (Value Object)", stats, "plainTraversalTimeMs", () => {
    let dummyCount = 0;
    for (let i = 0; i < NUM_ENTRIES; i++) {
      if (valueObjArray![i]) continue;
    }
    return dummyCount;
  });

  console.log("\n--- PROPERTY ACCESS ---");
  benchmarkStats("Property Access (Value Object)", stats, "propAccessTimeMs", () => {
    let sum = 0;
    for (let i = 0; i < NUM_ENTRIES; i++) {
      sum += valueObjArray![i].details.ag.a.s.r;
    }
    return sum;
  });

  console.log("\n--- FILTERING ---");
  benchmarkStats("Filtering (Value Object)", stats, "filterTimeMs", () => {
    let matched = 0;
    for (let i = 0; i < NUM_ENTRIES; i++) {
      if (valueObjArray![i].details.la > 1772455500) matched++;
    }
    return matched;
  });

  console.log("\n--- MUTATION ---");
  benchmarkStats("Mutation (Value Object)", stats, "mutationTimeMs", () => {
    for (let i = 0; i < NUM_ENTRIES; i++) {
      valueObjArray![i].details.la += 1;
    }
  });

  // Clear memory
  valueObjArray.length = 0;
  valueObjArray = null;
  Bun.gc(true);

  // Save Stats
  saveStats("stats.json", "value object", stats);
}

runBenchmark().catch(console.error);

File: language_construct_performance/python/scripts/idiomatic/value_obj_fixed_properties.py
import os
import gc
from dataclasses import dataclass
from typing import List, Optional, Any
import sys, os
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from benchmark_utils import NUM_ENTRIES, benchmark_stats, save_stats

# --- VALUE OBJECT CLASSES ---

@dataclass(slots=True)
class AgSType:
    t: int

@dataclass(slots=True)
class AgASubSType:
    t: int
    s: int
    r: int

@dataclass(slots=True)
class AgAType:
    t: int
    s: AgASubSType

@dataclass(slots=True)
class AgType:
    s: AgSType
    a: AgAType

@dataclass(slots=True)
class DetailsType:
    provider: str
    accountId: str
    principal: bool
    tags: List[str]
    mfas: str
    la: int
    s: int
    cpd: int
    pcb: str
    lld: int
    cd: int
    cb: str
    ub: str
    ud: int
    ua: int
    ut: int
    ag: AgType

@dataclass(slots=True)
class ValueObjectNode:
    label: str
    id: str
    accessTo: List[str]
    details: DetailsType
    edgeTo: List[str]

# --- FACTORIES ---

def create_value_object(index: int) -> ValueObjectNode:
    return ValueObjectNode(
        label=f'user-dev-test-{index}',
        id=f'u.{16406 + index}',
        accessTo=['s.[s3].UACDR', 'a.[s3].DARC', 's.[secretsmanager].RACDU', 's.[dynamodb].RCDAU'],
        details=DetailsType(
            provider='aws',
            accountId='568709751681',
            principal=True,
            tags=['aKIAYI2NaRQPOT', 'dev testing local'],
            mfas='',
            la=1772454942 + (index % 1000),
            ut=2,
            s=1,
            cpd=0,
            pcb='-',
            lld=0,
            cd=1763097939000,
            cb='-',
            ub='-',
            ud=0,
            ua=1772526871591,
            ag=AgType(
                s=AgSType(167),
                a=AgAType(
                    t=3187978,
                    s=AgASubSType(t=3187978, s=3149311, r=42506)
                )
            )
        ),
        edgeTo=['r.392', 'r.40', 'r.41', 'update', 'administrator', 'create', 'delete', 'read']
    )

# --- BENCHMARK RUNNER ---

def run_benchmark():
    stats = {}
    print(f"Starting Value Object Idiomatic Fixed Properties Benchmark with {NUM_ENTRIES:,} entries...\n")

    value_obj_list = [None] * NUM_ENTRIES

    print("\n--- CREATION ---")
    def creation():
        for i in range(NUM_ENTRIES):
            value_obj_list[i] = create_value_object(i)
    benchmark_stats("Value Object Idiomatic Creation", stats, "creationTimeMs", creation, track_memory=True)

    print("\n--- PLAIN TRAVERSAL ---")
    def traversal():
        dummy_count = 0
        for i in range(NUM_ENTRIES):
            if value_obj_list[i] is not None:
                dummy_count += 1
        return dummy_count
    benchmark_stats("Plain Traversal (Value Object Idiomatic)", stats, "plainTraversalTimeMs", traversal)

    print("\n--- PROPERTY ACCESS ---")
    def property_access():
        sum_val = 0
        for i in range(NUM_ENTRIES):
            sum_val += value_obj_list[i].details.ag.a.s.r
        return sum_val
    benchmark_stats("Property Access (Value Object Idiomatic)", stats, "propAccessTimeMs", property_access)

    print("\n--- FILTERING ---")
    def filtering():
        matched = 0
        for i in range(NUM_ENTRIES):
            if value_obj_list[i].details.la > 1772455500:
                matched += 1
        return matched
    benchmark_stats("Filtering (Value Object Idiomatic)", stats, "filterTimeMs", filtering)

    print("\n--- MUTATION ---")
    def mutation():
        for i in range(NUM_ENTRIES):
            value_obj_list[i].details.la += 1
    benchmark_stats("Mutation (Value Object Idiomatic)", stats, "mutationTimeMs", mutation)

    print("\n--- DELETE PROPERTY ---")
    def delete_property():
        for i in range(NUM_ENTRIES):
            try:
                del value_obj_list[i].details.ud
            except AttributeError:
                pass
    benchmark_stats("Delete Property (Value Object Idiomatic)", stats, "deletePropertyTimeMs", delete_property)

    # Clear memory
    value_obj_list.clear()
    gc.collect()

    # Save Stats
    save_stats("stats.json", "value object idiomatic", stats)

if __name__ == "__main__":
    try:
        run_benchmark()
    except Exception as e:
        print(f"Error: {e}")
        import sys
        sys.exit(1)
